src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc

```diff
@@ -1164,11 +1161,9 @@ static void grpc_resolve_address_ares_impl(const char* name,
                                            grpc_resolved_addresses** addrs) {
   grpc_resolve_address_ares_request* r =
       new grpc_resolve_address_ares_request();
+  absl::MutexLock lock(&r->mu);
```

Comment:
Another option here would be to use two-phase initialization for `grpc_ares_request`, so that we can create the object and store it in the `grpc_resolve_address_ares_request` object first, and then start the request.

The way you have this is fine for now, but please add a TODO to consider cleaning this up as part of the API changes we're making around DNS resolution.

Reply:
Definitely agree this two phase API would be cleaner - I've added a TODO